File: ios/Sources/SportsGreenMooveApp/GoogleMapsRoutePreview.swift
import SwiftUI

#if os(iOS) && canImport(GoogleMaps) && canImport(UIKit)
import CoreLocation
import GoogleMaps
import UIKit
#endif

struct GoogleMapsRoutePreviewCard: View {
    let ride: LiveRideSnapshot
    let preview: MapRoutePreview?

    var body: some View {
        ZStack(alignment: .bottomLeading) {
            content
            RoutePreviewOverlay(ride: ride)
        }
        .frame(height: 250)
        .frame(maxWidth: .infinity)
        .clipShape(RoundedRectangle(cornerRadius: SGMRadius.xl, style: .continuous))
        .background(SGM.bgCard, in: RoundedRectangle(cornerRadius: SGMRadius.xl, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: SGMRadius.xl, style: .continuous).stroke(SGM.border, lineWidth: 1))
    }

    @ViewBuilder
    private var content: some View {
        if !NativeGoogleMapsLifecycle.isSdkLinked {
            RoutePreviewState(
                title: "GOOGLE MAPS REQUIS",
                message: "Le package GoogleMaps doit être lié au projet iOS."
            )
        } else if !NativeGoogleMapsLifecycle.isConfigured {
            RoutePreviewState(
                title: "CLÉ GOOGLE MAPS REQUISE",
                message: "Ajoutez SGM_GOOGLE_MAPS_IOS_API_KEY pour afficher l'itinéraire natif."
            )
        } else if let preview {
            #if os(iOS) && canImport(GoogleMaps) && canImport(UIKit)
            NativeGoogleMapsRoutePreview(preview: preview)
            #else
            RoutePreviewState(
                title: "GOOGLE MAPS REQUIS",
                message: "Le SDK Google Maps iOS n'est pas disponible dans cette cible."
            )
            #endif
        } else {
            RoutePreviewState(
                title: "ITINÉRAIRE MANQUANT",
                message: "Le trajet actif ne contient pas encore de coordonnées départ-arrivée."
            )
        }
    }
}

private struct RoutePreviewState: View {
    let title: String
    let message: String

    var body: some View {
        VStack(spacing: 8) {
            Text(title)
                .font(.sgmDisplay(24))
                .tracking(.sgmWide(for: 24))
                .foregroundStyle(SGM.textOnGreen)
                .multilineTextAlignment(.center)
            Text(message)
                .font(.sgmBody(13, weight: .semibold))
                .foregroundStyle(SGM.textOnGreen.opacity(0.68))
                .multilineTextAlignment(.center)
        }
        .padding(.horizontal, 24)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(SGM.heroGradient)
        .overlay(SGMGridTexture())
    }
}

private struct RoutePreviewOverlay: View {
    let ride: LiveRideSnapshot

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text("SUIVI EN DIRECT")
                .font(.sgmBody(11, weight: .bold))
                .tracking(.sgmWider(for: 11))
                .foregroundStyle(SGM.greenLight)
            Text(ride.etaLabel)
                .font(.sgmBody(13, weight: .bold))
                .foregroundStyle(SGM.textOnGreen)
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 10)
        .background(SGM.heroStart.opacity(0.86), in: RoundedRectangle(cornerRadius: SGMRadius.lg, style: .continuous))
        .padding(14)
    }
}

#if os(iOS) && canImport(GoogleMaps) && canImport(UIKit)
private struct NativeGoogleMapsRoutePreview: UIViewRepresentable {
    let preview: MapRoutePreview

    func makeUIView(context: Context) -> GMSMapView {
        let center = preview.center
        let camera = GMSCameraPosition.camera(
            withLatitude: center.lat,
            longitude: center.lng,
            zoom: preview.routeZoom
        )
        let mapView = GMSMapView.map(withFrame: .zero, camera: camera)
        mapView.settings.compassButton = false
        mapView.settings.indoorPicker = false
        mapView.settings.myLocationButton = false
        mapView.settings.rotateGestures = false
        mapView.settings.tiltGestures = false
        mapView.settings.zoomGestures = true
        mapView.isTrafficEnabled = true
        return mapView
    }

    func updateUIView(_ mapView: GMSMapView, context: Context) {
        mapView.clear()
        let path = GMSMutablePath()
        MapRoutePolyline.points(for: preview).forEach { point in
            path.addLatitude(point.lat, longitude: point.lng)
        }

        let polyline = GMSPolyline(path: path)
        polyline.strokeWidth = 5
        polyline.strokeColor = UIColor(SGM.green)
        polyline.map = mapView

        let start = GMSMarker(position: CLLocationCoordinate2D(latitude: preview.start.lat, longitude: preview.start.lng))
        start.title = "Départ"
        start.map = mapView

        let end = GMSMarker(position: CLLocationCoordinate2D(latitude: preview.end.lat, longitude: preview.end.lng))
        end.title = "Arrivée"
        end.map = mapView

        let center = preview.center
        mapView.animate(to: GMSCameraPosition.camera(
            withLatitude: center.lat,
            longitude: center.lng,
            zoom: preview.routeZoom
        ))
    }
}

private extension MapRoutePreview {
    var center: MapPoint {
        MapPoint(lat: (start.lat + end.lat) / 2, lng: (start.lng + end.lng) / 2)
    }

    var routeZoom: Float {
        let delta = max(abs(start.lat - end.lat), abs(start.lng - end.lng))
        switch delta {
        case ..<0.015: return 14
        case ..<0.06: return 12
        case ..<0.18: return 10
        default: return 8
        }
    }
}
#endif

File: android/app/src/main/java/be/sportgreenmoove/app/ui/GoogleMapsRoutePreview.kt
package be.sportgreenmoove.app.ui

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.em
import androidx.compose.ui.unit.sp
import be.sportgreenmoove.app.R
import be.sportgreenmoove.app.data.LiveRideSnapshot
import be.sportgreenmoove.app.data.MapRoutePreview
import be.sportgreenmoove.app.design.Sgm
import be.sportgreenmoove.app.design.SgmColor
import be.sportgreenmoove.app.design.SgmType
import be.sportgreenmoove.app.domain.MapRoutePolyline
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.maps.android.compose.GoogleMap
import com.google.maps.android.compose.MapProperties
import com.google.maps.android.compose.MapUiSettings
import com.google.maps.android.compose.Marker
import com.google.maps.android.compose.MarkerState
import com.google.maps.android.compose.Polyline
import com.google.maps.android.compose.rememberCameraPositionState

@Composable
fun GoogleMapsRoutePreviewCard(
    ride: LiveRideSnapshot,
    preview: MapRoutePreview?,
) {
    val apiKey = stringResource(R.string.sgm_google_maps_android_api_key).trim()
    Box(
        modifier = Modifier
            .padding(horizontal = 20.dp)
            .fillMaxWidth()
            .height(250.dp)
            .sgmTestTag(SgmTestTags.GoogleMapsRoutePreview)
            .clip(RoundedCornerShape(22.dp))
            .background(Sgm.colors.bgCard)
            .border(BorderStroke(1.dp, Sgm.colors.border), RoundedCornerShape(22.dp)),
    ) {
        when {
            apiKey.isBlank() -> RoutePreviewState(
                title = "GOOGLE MAPS REQUIS",
                body = "Ajoutez SGM_GOOGLE_MAPS_ANDROID_API_KEY pour afficher l'itinéraire natif.",
            )

            preview == null -> RoutePreviewState(
                title = "ITINÉRAIRE MANQUANT",
                body = "Le trajet actif ne contient pas encore de coordonnées départ-arrivée.",
            )

            else -> NativeGoogleMap(preview = preview)
        }
        RoutePreviewOverlay(ride)
    }
}

@Composable
private fun NativeGoogleMap(preview: MapRoutePreview) {
    val points = remember(preview) {
        MapRoutePolyline.pointsFor(preview).map { LatLng(it.lat, it.lng) }
    }
    val start = LatLng(preview.start.lat, preview.start.lng)
    val end = LatLng(preview.end.lat, preview.end.lng)
    val center = LatLng(
        (preview.start.lat + preview.end.lat) / 2.0,
        (preview.start.lng + preview.end.lng) / 2.0,
    )
    val cameraState = rememberCameraPositionState {
        position = CameraPosition.fromLatLngZoom(center, routeZoom(preview))
    }

    GoogleMap(
        modifier = Modifier.fillMaxWidth().height(250.dp),
        cameraPositionState = cameraState,
        properties = MapProperties(isTrafficEnabled = true),
        uiSettings = MapUiSettings(
            compassEnabled = false,
            indoorLevelPickerEnabled = false,
            mapToolbarEnabled = false,
            myLocationButtonEnabled = false,
            rotationGesturesEnabled = false,
            tiltGesturesEnabled = false,
            zoomControlsEnabled = false,
        ),
    ) {
        Polyline(points = points, color = SgmColor.Green, width = 10f)
        Marker(state = MarkerState(position = start), title = "Départ")
        Marker(state = MarkerState(position = end), title = "Arrivée")
    }
}

@Composable
private fun RoutePreviewState(title: String, body: String) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(SgmColor.HeroGradient)
            .padding(horizontal = 24.dp),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text(
            title,
            style = SgmType.DisplayXL.copy(color = SgmColor.TextOnGreen, fontSize = 24.sp, letterSpacing = 0.06.em),
            textAlign = TextAlign.Center,
        )
        Text(
            body,
            style = SgmType.BodySM.copy(color = SgmColor.TextOnGreen.copy(alpha = 0.68f), fontSize = 13.sp),
            textAlign = TextAlign.Center,
        )
    }
}

@Composable
private fun BoxScope.RoutePreviewOverlay(ride: LiveRideSnapshot) {
    Column(
        modifier = Modifier
            .align(Alignment.BottomStart)
            .padding(14.dp)
            .clip(RoundedCornerShape(16.dp))
            .background(SgmColor.HeroStart.copy(alpha = 0.86f))
            .padding(horizontal = 14.dp, vertical = 10.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Text(
            "SUIVI EN DIRECT",
            style = SgmType.Eyebrow.copy(color = SgmColor.GreenLight, fontSize = 11.sp, letterSpacing = 0.14.em),
        )
        Text(
            ride.etaLabel,
            style = SgmType.BodySM.copy(color = SgmColor.TextOnGreen, fontSize = 13.sp, fontWeight = FontWeight.Bold),
        )
    }
}

private fun routeZoom(preview: MapRoutePreview): Float {
    val latDelta = kotlin.math.abs(preview.start.lat - preview.end.lat)
    val lngDelta = kotlin.math.abs(preview.start.lng - preview.end.lng)
    val delta = maxOf(latDelta, lngDelta)
    return when {
        delta < 0.015 -> 14f
        delta < 0.06 -> 12f
        delta < 0.18 -> 10f
        else -> 8f
    }
}
